File: clients/scala/src/main/scala/ai_cup_22/model/Item.scala
package ai_cup_22.model

import ai_cup_22.util.StreamUtil

/**
 * Lootable item
 */
sealed trait Item {
    /**
     * Write Item to output stream
     */
    def writeTo(stream: java.io.OutputStream): scala.Unit
}

object Item {
    /**
     * Weapon
     *
     * @param typeIndex Weapon type index (starting with 0)
     */
    case class Weapon(typeIndex: Int) extends Item {
        /**
         * Write Weapon to output stream
         */
        override def writeTo(stream: java.io.OutputStream): scala.Unit = {
            StreamUtil.writeInt(stream, Weapon.TAG)
            StreamUtil.writeInt(stream, typeIndex)
        }
    
        /**
         * Get string representation of Weapon
         */
        override def toString(): String = {
            var stringBuilder = new StringBuilder("Weapon { ")
            stringBuilder.append("typeIndex: ")
            stringBuilder.append(typeIndex)
            stringBuilder.append(" }")
            stringBuilder.toString()
        }
    }
    
    object Weapon {
        val TAG: Int = 0
    
        /**
         * Read Weapon from input stream
         */
        def readFrom(stream: java.io.InputStream): Weapon = Weapon(
            StreamUtil.readInt(stream)
        )
    }

    /**
     * Shield potions
     *
     * @param amount Amount of potions
     */
    case class ShieldPotions(amount: Int) extends Item {
        /**
         * Write ShieldPotions to output stream
         */
        override def writeTo(stream: java.io.OutputStream): scala.Unit = {
            StreamUtil.writeInt(stream, ShieldPotions.TAG)
            StreamUtil.writeInt(stream, amount)
        }
    
        /**
         * Get string representation of ShieldPotions
         */
        override def toString(): String = {
            var stringBuilder = new StringBuilder("ShieldPotions { ")
            stringBuilder.append("amount: ")
            stringBuilder.append(amount)
            stringBuilder.append(" }")
            stringBuilder.toString()
        }
    }
    
    object ShieldPotions {
        val TAG: Int = 1
    
        /**
         * Read ShieldPotions from input stream
         */
        def readFrom(stream: java.io.InputStream): ShieldPotions = ShieldPotions(
            StreamUtil.readInt(stream)
        )
    }

    /**
     * Ammo
     *
     * @param weaponTypeIndex Weapon type index (starting with 0)
     * @param amount Amount of ammo
     */
    case class Ammo(weaponTypeIndex: Int, amount: Int) extends Item {
        /**
         * Write Ammo to output stream
         */
        override def writeTo(stream: java.io.OutputStream): scala.Unit = {
            StreamUtil.writeInt(stream, Ammo.TAG)
            StreamUtil.writeInt(stream, weaponTypeIndex)
            StreamUtil.writeInt(stream, amount)
        }
    
        /**
         * Get string representation of Ammo
         */
        override def toString(): String = {
            var stringBuilder = new StringBuilder("Ammo { ")
            stringBuilder.append("weaponTypeIndex: ")
            stringBuilder.append(weaponTypeIndex)
            stringBuilder.append(", ")
            stringBuilder.append("amount: ")
            stringBuilder.append(amount)
            stringBuilder.append(" }")
            stringBuilder.toString()
        }
    }
    
    object Ammo {
        val TAG: Int = 2
    
        /**
         * Read Ammo from input stream
         */
        def readFrom(stream: java.io.InputStream): Ammo = Ammo(
            StreamUtil.readInt(stream),
            StreamUtil.readInt(stream)
        )
    }

    /**
     * Read Item from input stream
     */
    def readFrom(stream: java.io.InputStream): Item = {
        StreamUtil.readInt(stream) match {
            case Weapon.TAG => Weapon.readFrom(stream)
            case ShieldPotions.TAG => ShieldPotions.readFrom(stream)
            case Ammo.TAG => Ammo.readFrom(stream)
            case _ => throw new java.io.IOException("Unexpected tag value")
        }
    }
}

File: clients/swift/Sources/AiCup22/Model/Item.swift
/// Lootable item
enum Item {
    /// Weapon
    ///
    /// - typeIndex: Weapon type index (starting with 0)
    case weapon(typeIndex: Int32)

    /// Shield potions
    ///
    /// - amount: Amount of potions
    case shieldPotions(amount: Int32)

    /// Ammo
    ///
    /// - weaponTypeIndex: Weapon type index (starting with 0)
    /// - amount: Amount of ammo
    case ammo(weaponTypeIndex: Int32, amount: Int32)

    /// Read Item from input stream
    static func readFrom<S: InputStream>(_ stream: S) -> Item {
        switch stream.readInt32() {
            case 0:
                var typeIndex: Int32
                typeIndex = stream.readInt32()
                return Item.weapon(typeIndex: typeIndex)
            case 1:
                var amount: Int32
                amount = stream.readInt32()
                return Item.shieldPotions(amount: amount)
            case 2:
                var weaponTypeIndex: Int32
                weaponTypeIndex = stream.readInt32()
                var amount: Int32
                amount = stream.readInt32()
                return Item.ammo(weaponTypeIndex: weaponTypeIndex, amount: amount)
            default:
                fatalError("Unexpected tag value")
        }
    }

    /// Write Item to output stream
    func writeTo<S: OutputStream>(_ stream: S) {
        switch self {
            case let .weapon(typeIndex):
                stream.writeInt32(0)
                stream.writeInt32(typeIndex)
            case let .shieldPotions(amount):
                stream.writeInt32(1)
                stream.writeInt32(amount)
            case let .ammo(weaponTypeIndex, amount):
                stream.writeInt32(2)
                stream.writeInt32(weaponTypeIndex)
                stream.writeInt32(amount)
        }
    }
}
